lib/luahid.c

```diff
@@ -50,6 +46,45 @@ static const lunatik_class_t luahid_class = {
 	.sleep = true,
 };
 
+static const struct hid_device_id *luahid_setidtable(lua_State *L, int idx)
+{
+	if (lua_getfield(L, idx, "id_table") != LUA_TTABLE) /* id_table */
+		goto err;
+
+	size_t len = luaL_len(L, -1);
+	if (len == 0)
+		goto err;
+
+	struct hid_device_id *user_table = lunatik_checkalloc(L, sizeof(struct hid_device_id) * (len + 1));
+
+	for (size_t i = 0; i < len; i++) {
+		if (lua_geti(L, -1, i + 1) != LUA_TTABLE) { /* i-th entry of the table */
+			/*
+			 * lunatik_free(user_table);
+			 * luaL_error(L, "id_table entry #%zu is not a table", i + 1);
+			 */
+			goto err;
+		}
+
+		struct hid_device_id *cur_id = &user_table[i];
+		lunatik_optinteger(L, 1, cur_id, bus, HID_BUS_ANY);
+		lunatik_optinteger(L, 1, cur_id, group, HID_GROUP_ANY);
+		lunatik_optinteger(L, 1, cur_id, vendor, HID_ANY_ID);
+		lunatik_optinteger(L, 1, cur_id, product, HID_ANY_ID);
+		lunatik_optinteger(L, 1, cur_id, driver_data, 0);
+
+		lua_pop(L, 1); /* table entry */
+	}
+
+	memset(&user_table[len], 0, sizeof(struct hid_device_id));
+	lua_pop(L, 1); /* id_table */
+
+	return user_table;
```

Comment:
Why not initializing it with NULL, so we can avoid the last two lines, avoiding unnecessarily replicating code